[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/efurlanm/418/blob/master/hands-on-B04-euler.ipynb)

# Euler's Method

*Last edited 2023-06-16*

Euler's method, a numerical method for approximating solutions of first-order Initial Value Problems (IVP) used to solve ordinary differential equations (ODEs) with a given initial value, is one of the most basic explicit methods for numerical integration of ODEs, and was first proposed in the book Institutionum Calculi Integralis published in 1768–1870. As it is a first-order method, the local error (error per step) is proportional to the square of the step size, and the global error (error at a given time) is proportional to the step size. We can apply it to the approximate solution of an IVP $y'=f(x,y) \ , \ y(x_0)=y_0$ . As we know $x_0$ and $y_0 = y(x_0)$ , so we know how to calculate $y'(x_0)=f(x_0,y_0)$ . Thus, the line passing through $(x_0, y_0)$ with angular coefficient $y'(x_0),r_0(x)$ is given by $r_0(x)=y(x_0)+(x-x_0)y'(x_0)$ for step size $h=x_{k+1}-x_k \ , \ y(x_1) \approx y_1 = r_0(x_1) = y _0+hy'(x_0)$, that is, $y_1=y_0+hf (x_0,y_0)$ .

# Implementation

In [2]:
import numpy as np
from numpy import arange

np.set_printoptions(precision=4, suppress=True)

In [3]:
def euler(x, y, h, f, n):
    print(" n    x      y          d       y_n+1")
    print("--   ---   -------   -------   -------")
    for i in range(n):
        d = f(x, y)
        y2 = y + h * d
        print(f"{i:2}   {x:.1f}   {y:07.4f}   {d:07.4f}   {y2:07.4f}")
        x = x + h
        y = y2

# Example 1

<br>

Example from: Devin Henson. Euler's Method - Example 1. https://youtu.be/PwuZ3nir7d4 

$ \frac{dy}{dx} = x + 2y \ , \quad h = 0.1 $

Approximate $y(0.4)$ using Euler's method.

$X_n = X_{n-1} + h \\
Y_n = Y_{n-1} + h \cdot f(X_{n-1}, Y_{n-1})$

In [4]:
def f(x, y):
    return x + 2*y

x = 0.0
y = 0.0
h = 0.1
n = 5

euler(x, y, h, f, n)

 n    x      y          d       y_n+1
--   ---   -------   -------   -------
 0   0.0   00.0000   00.0000   00.0000
 1   0.1   00.0000   00.1000   00.0100
 2   0.2   00.0100   00.2200   00.0320
 3   0.3   00.0320   00.3640   00.0684
 4   0.4   00.0684   00.5368   00.1221


Result: the approximation to $y(0.4)$ is 0.0684 .

# Example 2

Based on: Khan Academy. Euler's method. https://youtu.be/q87L9R9v274

$ \frac{dy}{dx} = y \ , \quad y(0) = 1 $

The manual analytical solution is $y=e^x$

In [12]:
def f(x, y):
    return y

x = 0.0
y = 1.0
h = 1.0
n = 4

euler(x, y, h, f, n)

 n    x      y          d       y_n+1
--   ---   -------   -------   -------
 0   0.0   01.0000   01.0000   02.0000
 1   1.0   02.0000   02.0000   04.0000
 2   2.0   04.0000   04.0000   08.0000
 3   3.0   08.0000   08.0000   16.0000


Comparing with the manual analytical solution of $e^3 \approx 20$ for $x=3$, we can see that for $h=1.0$ the result of 8.0 was far from expected.

Let's now reduce the value of $h$ :

In [18]:
def f(x, y):
    return y

x = 0.0
y = 1.0
h = 0.1
n = 31

euler(x, y, h, f, n)

 n    x      y          d       y_n+1
--   ---   -------   -------   -------
 0   0.0   01.0000   01.0000   01.1000
 1   0.1   01.1000   01.1000   01.2100
 2   0.2   01.2100   01.2100   01.3310
 3   0.3   01.3310   01.3310   01.4641
 4   0.4   01.4641   01.4641   01.6105
 5   0.5   01.6105   01.6105   01.7716
 6   0.6   01.7716   01.7716   01.9487
 7   0.7   01.9487   01.9487   02.1436
 8   0.8   02.1436   02.1436   02.3579
 9   0.9   02.3579   02.3579   02.5937
10   1.0   02.5937   02.5937   02.8531
11   1.1   02.8531   02.8531   03.1384
12   1.2   03.1384   03.1384   03.4523
13   1.3   03.4523   03.4523   03.7975
14   1.4   03.7975   03.7975   04.1772
15   1.5   04.1772   04.1772   04.5950
16   1.6   04.5950   04.5950   05.0545
17   1.7   05.0545   05.0545   05.5599
18   1.8   05.5599   05.5599   06.1159
19   1.9   06.1159   06.1159   06.7275
20   2.0   06.7275   06.7275   07.4002
21   2.1   07.4002   07.4002   08.1403
22   2.2   08.1403   08.1403   08.9543
23   2.3   08.9543   08.95

With $h=0.1$ the result of 17 was closer to the expected 20.

# Example 3

<br>

Based on: The Organic Chemistry Tutor. Euler's Method. https://youtu.be/ukNbG7muKho

$ \frac{dy}{dx} = x + 2y \ , \ h=0.1 \ , \ y(2)=3 \ , \ y(2.5)=? $

In [19]:
def f(x, y):
    return x + 2*y

x = 2.0
y = 3.0
h = 0.1
n = 6

euler(x, y, h, f, n)

 n    x      y          d       y_n+1
--   ---   -------   -------   -------
 0   2.0   03.0000   08.0000   03.8000
 1   2.1   03.8000   09.7000   04.7700
 2   2.2   04.7700   11.7400   05.9440
 3   2.3   05.9440   14.1880   07.3628
 4   2.4   07.3628   17.1256   09.0754
 5   2.5   09.0754   20.6507   11.1404


Result: $ y(2.5)= 9.0754 $

# Example 4

Based on: Jenn from Calcworkshop. Euler's Method. https://youtu.be/8cW_CQ77ayI

Use Euler's method with $h=0.1$ to approximate $y(0.3)$ for the IVP $ \frac{dy}{dx} = x + y \ , \quad y(0)=1 $ .

In [20]:
def f(x, y):
    return x + y

x = 0.0
y = 1.0
h = 0.1
n = 4

euler(x, y, h, f, n)

 n    x      y          d       y_n+1
--   ---   -------   -------   -------
 0   0.0   01.0000   01.0000   01.1000
 1   0.1   01.1000   01.2000   01.2200
 2   0.2   01.2200   01.4200   01.3620
 3   0.3   01.3620   01.6620   01.5282


The result of approximation $y(0.3)$ is 1.362 .

# Conclusion

This work sought to show an introduction to the numerical method for approximating solutions of first order Initial Value Problems (IVP) used to solve ordinary differential equations (ODEs) with a given initial value. An implementation was made and some examples showed its application. In particular, example 2 showed that a large number of iterations may be necessary to reduce the error and reach a result close to the expected one.

# References

EULER’S METHOD. DIFFERENTIAL EQUATIONS. [S. l.: s. n.], 2014. Available at: https://www.youtube.com/watch?v=q87L9R9v274. Accessed on: 10 Jun. 2023.

FIREFLY LECTURES. Euler’s Method - Example 1. [S. l.: s. n.], 28 Apr. 2013. Available at: https://www.youtube.com/watch?v=PwuZ3nir7d4. Accessed on: 10 Jun. 2023.

JENN. Euler’s Method - A Simple Table That Works Every Time - YouTube. [S. l.: s. n.], 2016. Available at: https://www.youtube.com/watch?v=8cW_CQ77ayI. Accessed on: 10 Jun. 2023.

RUGGIERO, M. A. G.; LOPES, V. L. D. R. Cálculo numérico: aspectos teóricos e computacionais. [S. l.]: Pearson Makron Books, 1996. vol. 1, . Available at: https://worldcat.org/en/title/1042351138.

THE ORGANIC CHEMISTRY TUTOR. Euler’s Method Differential Equations, Examples, Numerical Methods, Calculus. [S. l.: s. n.], 11 Feb. 2017. Available at: https://www.youtube.com/watch?v=ukNbG7muKho. Accessed on: 10 Jun. 2023.
